lib/util/semver.js

```diff
@@ -0,0 +1,483 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+
+"use strict";
+
+/** @typedef {(string|number|undefined|[])[]} SemVerRange */
+/** @typedef {(string|number|undefined|[])[]} SemVerVersion */
+
+const removeStars = arr => {
+	while (/^[Xx*]$/.test(arr[arr.length - 1])) arr.pop();
+	return arr;
+};
+
+/**
+ * @param {string} str version string
+ * @returns {SemVerVersion} parsed version
+ */
+const parseVersion = str => {
+	if (!str) return [];
+	const match = /^([^-+]+)(?:-([^+]+))?(?:\+(.+))?$/.exec(str);
+	const convertToNumber = item => (`${+item}` === item ? +item : item);
+	const version = removeStars(match[1].split(".").map(convertToNumber));
+	const prerelease = match[2]
+		? removeStars(match[2].split(".").map(convertToNumber))
+		: [];
+	const build = match[3] ? match[3].split(".").map(convertToNumber) : [];
+	if (prerelease.length) {
+		return build.length
+			? // eslint-disable-next-line no-sparse-arrays
+			  [...version, , ...prerelease, [], ...build]
+			: // eslint-disable-next-line no-sparse-arrays
+			  [...version, , ...prerelease];
+	} else {
+		// eslint-disable-next-line no-sparse-arrays
+		return build.length ? [...version, [], ...build] : version;
+	}
+};
+exports.parseVersion = parseVersion;
+
+/* eslint-disable eqeqeq */
+const versionToString = version => {
+	if (!version) return "[unknown]";
+	for (var str = "", needDot = 1, i = 0; i < version.length; i++) {
+		var item = version[i];
+		var t = (typeof item)[0];
+		needDot--;
+		str +=
+			t == "u"
+				? // undefined: prerelease marker, add an "-"
+				  "-"
+				: t == "o"
+				? // object: build marker, add an "+"
+				  "+"
+				: // number or string: add the item, set flag to add an "." between two of them
+				  (needDot > 0 ? "." : "") + ((needDot = 2), item);
+	}
+	return str;
+};
+/* eslint-enable eqeqeq */
+exports.versionToString = versionToString;
+
+// must be a minimized version of above
+exports.versionToStringRuntimeCode = runtimeTemplate =>
+	`var versionToString = ${runtimeTemplate.basicFunction("version", [
+		"// see webpack/lib/util/semver.js for original code",
+		'if(!version)return"[unknown]";for(var r="",n=1,o=0;o<version.length;o++){var e=version[o],t=(typeof e)[0];n--,r+="u"==t?"-":"o"==t?"+":(n>0?".":"")+(n=2,e)}return r'
```

Comment:
```suggestion
	`var versionToString = ${runtimeTemplate.basicFunction("v", [
		"// see webpack/lib/util/semver.js for original code",
		'if(!v)return"[unknown]";for(var r="",n=1,o=0;o<v.length;o++){var e=v[o],t=(typeof e)[0];n--,r+="u"==t?"-":"o"==t?"+":(n>0?".":"")+(n=2,e)}return r'
```

Reply:
This is not needed. The code runs through a minimizier anyway. Maybe we should switch off mangling for this at all.